Source/WebCore/html/canvas/GPUCanvasContextCocoa.cpp

```diff
@@ -188,10 +205,16 @@ RefPtr<GraphicsLayerContentsDisplayDelegate> GPUCanvasContextCocoa::layerContent
 
 void GPUCanvasContextCocoa::prepareForDisplay()
 {
-    m_presentationContext->prepareForDisplay([protectedThis = Ref { *this }] (auto sendRight) {
-        protectedThis->m_layerContentsDisplayDelegate->setDisplayBuffer(WTFMove(sendRight));
-        protectedThis->m_compositingResultsNeedsUpdating = false;
-    });
+    if (!isConfigured())
+        return;
+
+    ASSERT(m_configuration->frameCount < m_configuration->renderBuffers.size());
+
+    m_compositorIntegration->prepareForDisplay();
+
+    m_layerContentsDisplayDelegate->setDisplayBuffer(m_configuration->renderBuffers[m_configuration->frameCount]);
+    m_compositingResultsNeedsUpdating = false;
+    m_configuration->frameCount = (m_configuration->frameCount + 1) % m_configuration->renderBuffers.size();
```

Comment:
Is this always correct? What if in the case of triple buffering when rendering frame N, the render for frame N-2 is complete prior to the frame for render N-1? Shouldn't we choose the first available frame instead of round-robining?

Reply:
Isn't it bad to show frames out of order to the user?

Reply:
Do you think you could describe the frame ordering strategy you have in mind in more detail?

Reply:
We should always present the frames to the user in order but this is independent of the order we render to the triple buffers, I can draw an image

Reply:
I'm going to contradict what I wrote earlier. I think the issue is in the case of triple buffering, won't this mean when we display frame N, then we will render to frame N+2. It adds a latency of one frame

Reply:
Upon further discussion, I've filed https://bugs.webkit.org/show_bug.cgi?id=251625 and we're going to investigate this asynchronously.